src/well_known_messages.cc

```diff
@@ -0,0 +1,248 @@
+            dataValue->_value.Reserve(totalElements);
+            auto dest = dataValue->_value.AddNAlreadyReserved(totalElements);
+            memcpy(dest, arrayData, totalElements * sizeof(double));
+        }
+
+        //---------------------------------------------------------------------
+        //---------------------------------------------------------------------
+        const std::string& String2DArray::GetMessageName()
+        {
+            static const std::string messageName = "ni_protobuf_types_String2DArray";
+            return messageName;
+        }
+
+        //---------------------------------------------------------------------
+        //---------------------------------------------------------------------
+        const std::string& String2DArray::GetMessageUrl()
+        {
+            static const std::string messageUrl = "ni.protobuf.types.String2DArray";
+            return messageUrl;
+        }
+
+        //---------------------------------------------------------------------
+        //---------------------------------------------------------------------
+        void String2DArray::CopyArrayFromMessageToCluster(int rows, int columns, const std::shared_ptr<const LVMessageValue>& dataFieldValue, int8_t* start)
+        {
+            auto elementCount = rows * columns;
+            NumericArrayResize(GetTypeCodeForSize(sizeof(char*)), 2, start, elementCount);
+            auto array = *(LV2DArrayHandle*)start;
+            (*array)->dimensionSizes[0] = rows;
+            (*array)->dimensionSizes[1] = columns;
+
+            // Protect against a malformed message where the amount of data sent doesn't match the dimension sizes.
+            // LV will automatically pad/handle writing less data than was allocated to the array so we just need
+            // to make sure we don't write more data than was allocated to the array.
+            auto dataValue = std::static_pointer_cast<const LVRepeatedMessageValue<std::string>>(dataFieldValue);
+            auto stringElements = dataValue->_value.size();
+            auto stringsToCopy = std::min(elementCount, stringElements);
+            auto lvString = (*array)->bytes<LStrHandle>();
+            for (int i = 0; i < stringsToCopy; i++)
+            {
+                *lvString = nullptr;
+                SetLVString(lvString, dataValue->_value[i]);
+                lvString += 1;
+            }
+        }
+
+        void String2DArray::CopyArrayFromClusterToMessage(int totalElements, LV2DArrayHandle array, const std::shared_ptr<LVMessageValue>& dataFieldValue)
+        {
+            auto lvStr = (*array)->bytes<LStrHandle>();
+            auto dataValue = std::static_pointer_cast<LVRepeatedMessageValue<std::string>>(dataFieldValue);
+            for (int i = 0; i < totalElements; i++)
+            {
+                auto str = GetLVString(*lvStr);
+                dataValue->_value.Add(str);
+                lvStr += 1;
+            }
```

Comment:
I think you can avoid some unnecessary copies here.

```suggestion
            auto dataValue = std::static_pointer_cast<LVRepeatedMessageValue<std::string>>(dataFieldValue);
            dataValue->_value.Reserve(totalElements);
            for (int i = 0; i < totalElements; i++)
            {
                auto str = GetLVString(*lvStr);
                dataValue->_value.AddAlreadyReserved(std::move(str));
                lvStr += 1;
            }
```

If `Add(std::move(str))` doesn't compile then I think our protobuf library is outdated. The docs in https://protobuf.dev/reference/cpp/api-docs/google.protobuf.repeated_field/ say `Add` takes `const Element&` but the code in https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/repeated_field.h takes `Element` by value in order to support move semantics.

If we are using an outdated protobuf then this might work:
`*dataValue->_value.AddAlreadyReserved() = std::move(str);`

or you can resign yourself to making unnecessary string copies. At least do the reserve, though.

Reply:
Seems to work fine. Thanks for the suggestion. I also updated similar code from `cluster_copier.cc`.